### PR TITLE
@gnilekaw => Implement ASCII folding for partners' location search

### PR DESCRIPTION
**Pull request by anandaroop**
fold

Currently the location typeahead dropdown on /galleries will fail for cities that contain diacritics such as **Zürich** or **Bogotá** unless the user keys in the _exact_ accented letter.

Since these typeaheads are matching on the client side (rather than in Elasticsearch) we don't get any [ASCII folding behavior](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-asciifolding-tokenfilter.html) for free. Rather we have to provide it ourselves.

This PR implements the required ASCII folding by adding a small util class and a custom character map.

I looked into https://github.com/mplatt/fold-to-ascii for the character map but decided that our use case is too limited to justify adding a 25k library which essentially recreates all of Lucene's ASCII folding in Javascript. (But we should consider that in the future if we ever need it.)

Meanwhile I used the [character map](https://github.com/artsy/force/blob/master/components/commercial_filter/filters/location/character_map.coffee) that @mzikherman created for the /collect page for inspiration, and added to it, augmenting it with several new characters that we need for our current list of cities.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please assign someone to your PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by mzikherman** [comment]
😎

**Comment by starsirius** [comment]
😎

**Comment by joeyAghion** [comment]
> (But we should consider that in the future if we ever need it.)

We could also add support for an `?indexes[]=City` parameter to the `match/suggest` API endpoint, much like we do for the `match` endpoint.

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please assign someone to your PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by anandaroop** [comment]
Thanks @joey, good idea. Using ES autocompletes for that menu, [now that it's possible](https://github.com/artsy/gravity/pull/10591) 🎉 , would probably be a wise replacement for most if not all of the city-related functionality on Force.

I'd still be fine with merging this for now, to deal with the present Zürich problem, as reported on Slack.

File: apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee [hunk 1]
```diff
@@ -35,9 +36,10 @@ module.exports = class PartnerFilterFacet extends Backbone.Model
     escape= (s) ->
       s.replace /[-\/\\^$*+?.()|[\]{}]/g, '\\$&'
 
+    query = AsciiFolder.fold(query)
```
Comment (gnilekaw) [inline apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee]: #superminor Instead of overwriting `query`, one option would be to declare a new variable `var foldedQuery = `.  I've found this can be easier to read in the future.

File: apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee [hunk 2]
```diff
@@ -35,9 +36,10 @@ module.exports = class PartnerFilterFacet extends Backbone.Model
...
     regex = _s.clean(escape(query)).replace(' ', '\\W* \\W*')
     substrRegex = new RegExp(regex, 'i')
-    substrRegex.test string
+    substrRegex.test AsciiFolder.fold(string)
```
Comment (gnilekaw) [inline apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee]: Why is it necessary to fold twice?

File: apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee (hunk 1 above)
Comment (anandaroop) [inline apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee]: Good point, I've pushed up an update for that!

File: apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee (hunk 2 above)
Comment (anandaroop) [inline apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee]: Good question -- we fold both the query term _and_ the city whose name we are searching against, so that we are doing an apples-to-apples comparison of two folded strings.

Folding the city name alone might work most of the time, but would then fail for a user who actually _does_ key in the diacritic.

File: apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee [hunk 3]
```diff
@@ -35,9 +36,10 @@ module.exports = class PartnerFilterFacet extends Backbone.Model
     escape= (s) ->
       s.replace /[-\/\\^$*+?.()|[\]{}]/g, '\\$&'
 
-    regex = _s.clean(escape(query)).replace(' ', '\\W* \\W*')
+    foldedQuery = AsciiFolder.fold(query)
+    regex = _s.clean(escape(foldedQuery)).replace(' ', '\\W* \\W*')
     substrRegex = new RegExp(regex, 'i')
-    substrRegex.test string
+    substrRegex.test AsciiFolder.fold(string)
```
Comment (gnilekaw) [inline apps/galleries_institutions/components/filter_facet/partner_filter_facet.coffee]: I don't think you need to call fold twice maybe. I think you could delete `foldedQuery = AsciiFolder.fold(query)` all together.

```js
regex = _s.clean(escape(query)).replace(' ', '\\W* \\W*')
substrRegex = new RegExp(regex, 'i')
substrRegex.test AsciiFolder.fold(string)
```

File: components/util/ascii_folder.coffee [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  fold: (string) ->
+    folded = (@foldedCharacter(char) for char in string).join ''
+
+  foldedCharacter: (char) ->
+    @characterMap[char] or char
+
+  characterMap: 
+    'À': 'A'
+    'Á': 'A'
+    'Â': 'A'
```
Comment (mzikherman) [inline components/util/ascii_folder.coffee]: Sure 👍

**Comment by anandaroop** [comment]
Plot twist: https://github.com/artsy/force/pull/476

**tldr;** turns out ascii folding was already hiding out in one of our many dependencies, so let's use that instead of DIYing it

I'll close this PR in favor of that one.
